monai/handlers/checkpoint_saver.py

```diff
@@ -0,0 +1,194 @@
+        file_prefix (str): prefix for the filenames to which objects will be saved.
+        save_final (bool): whether to save checkpoint or session at final iteration or exception.
+        save_key_metric (bool): whether to save checkpoint or session according to key metric value.
+        key_metric_name (str): the name of key_metric in ignite metrics dictionary.
+            if None, use `engine.state.key_metric` instead.
+        key_metric_n_saved (int): save top N checkpoints or sessions sorted by the key metric.
+        epoch_save_interval (int): save checkpoint or session every N epochs.
+        epoch_n_saved (int): save latest N epochs' checkpoints or sessions, 'None' is to save all.
+        iteration_save_interval (int): save checkpoint or session every N iterations.
+        iteration_n_saved (int): save latest N iterations' checkpoints or sessions, 'None' is to save all.
+        convert_to_single_gpu (bool): if current model is trained by multi GPU and
+            want to save to single GPU model, need to save `object.module` instead.
+
+    Note:
+        CheckpointHandler can be used during training, validation or evaluation.
+        example of saved files:
+            checkpoint_iteration=400.pth
+            checkpoint_iteration=800.pth
+            checkpoint_epoch=1.pth
+            checkpoint_final_iteration=1000.pth
+            checkpoint_key_metric=0.9387.pth
+
+    """
+
+    def __init__(
+        self,
+        save_dir,
+        save_dict,
+        name=None,
+        file_prefix="",
+        save_final=False,
+        save_key_metric=False,
+        key_metric_name=None,
+        key_metric_n_saved=1,
+        epoch_save_interval=0,
+        epoch_n_saved=None,
+        iteration_save_interval=0,
+        iteration_n_saved=None,
+        convert_to_single_gpu=False,
+    ):
+        assert save_dir is not None, "must provide directory to save the checkpoints."
+        self.save_dir = save_dir
+        assert save_dict is not None and len(save_dict) > 0, "must provide source objects to save."
+        if convert_to_single_gpu:
+            for k, v in save_dict.items():
+                if hasattr(v, "module"):
+                    save_dict[k] = v.module
+        self.save_dict = save_dict
+        self.logger = None if name is None else logging.getLogger(name)
+        self.save_final = save_final
+        self.save_key_metric = save_key_metric
+        self.key_metric_name = key_metric_name
+        self.key_metric_n_saved = key_metric_n_saved
+        self.epoch_save_interval = epoch_save_interval
+        self.epoch_n_saved = epoch_n_saved
+        self.iteration_save_interval = iteration_save_interval
+        self.iteration_n_saved = iteration_n_saved
+
+        if self.save_final is True:
+
+            def _final_func(engine):
+                return engine.state.iteration
+
+            self.final_checkpoint = ModelCheckpoint(
```

Comment:
Do we need to have multiple `ModelCheckpoint` objects? I would think that we can attach the same object to multiple events with the `every` filter used (eg. `engine.add_event_handler(Events.ITERATION_COMPLETED(every=n_iter)`) and not have to provide event methods ourselves. The way I did it in the past was have just a function which had an event handler for attaching a `ModelCheckpoint` instance when the run starts. It should be possible to do that with modifications to attach it to save at the end of iterarions and on exception raises. 

One other thing it did was attach a `logging.FileHandler` to the engine's logger to spit out the log to a file in the directory with the .pth files.

Reply:
I think it's nice to have an independent `ModelCheckpoint` to save the top N best models, the other checkpoint could merge.

Reply:
Hi @ericspod ,

Thanks for your review and comments.
Do you mean to use only 1 `ModelCheckpoint ` and silently replace `self._score_name` and `self._score_function` every time we trigger an event-handler?
I feel it's maybe not the recommended way to use this module as these 2 members are protected with "_".
And actually, even I used several `self._score_function` instances in this handler, it's very light an object, not some memory-consuming objects I think.
What do you think?

Reply:
The scores are different for saving iteration or epoch, but I'd think we'd be doing only one or the other. That would reduce the number of `ModelCheckpoint` objects by one, but you can then in the same way combine that with the `key_metric_checkpoint` and just choose different values for the arguments. `final_checkpoint` is separate  because you want to have different filenames but one could define the event handler to copy the generated .pth file to a new filename instead. I'm just thinking that we can get the result we want with only one `ModelCheckpoint` object that gets attached in a function that creates it rather than creating a new class.

Reply:
OK, I see. Let me try to reduce the `ModelCheckpoint ` objects, as @wyli said, may keep `best_metric` separately.

Reply:
Hi @ericspod ,

Usually, we save the best metric model in validator engine instead of trainer engine, but we may also save both the models based on `train_metrics`(like train accuracy or train loss) and `epoch` or `iteration`, I think we can't merge `key_metric_checkpoint ` with `epoch` and `iteration` checkpoint.
I reduced `ModelCheckpoint` to only keep `final`, `key_metric` and `interval` now.
Thanks.